tests/ops/qubit/test_attributes.py

```diff
@@ -471,3 +472,61 @@ def test_pcphase(self):
         expected_mat = np.array(mats)
         assert np.allclose(mat1, expected_mat)
         assert np.allclose(mat2, expected_mat)
+
+
+all_qubit_operators = sorted(qml.ops.qubit.__all__)  # pylint: disable=no-member
+unitarily_generated_ops = sorted(list(has_unitary_generator))
+
+
+class TestHasUnitaryGenerator:
+    """Test that all operations in the ``has_unitary_generator`` attribute
+    actually have unitary generators."""
+
+    @pytest.mark.parametrize("entry", unitarily_generated_ops)
+    def test_generator_unitarity(self, entry):
+        """Test directly that generators of the operators in the ``has_unitary_generator``
+        attribute are unitary up to a factor of 2."""
+        op_class = getattr(qml, entry)
+        phi = 1.23
+        wires = [0, 1, 2] if op_class.num_wires is AnyWires else list(range(op_class.num_wires))
+        if op_class is qml.PauliRot:
+            op = op_class(phi, pauli_word="XYZ", wires=wires)  # PauliRot has num_wires == AnyWires
+        elif op_class is qml.PCPhase:
+            op = op_class(phi, dim=(2 ** len(wires) - 1), wires=wires)
+        else:
+            op = op_class(phi, wires=wires)
+        gen = qml.generator(op, format="observable")
+        assert not qml.math.isclose(
+            qml.matrix(qml.prod(qml.adjoint(gen), gen))[0, 0], 0
+        )  # Not unitary if diagonal contains zeros
+        # Some generators are unitary up to a factor - in this case product
+        # of generator with its adjoint has square of this factor on
+        # diagonal
+        coeff = 1.0 / np.sqrt(qml.matrix(qml.prod(qml.adjoint(gen), gen))[0, 0])
+        assert qml.is_unitary(qml.s_prod(coeff, gen))
```

Comment:
Thanks for your help, this way is a lot clearer! Because `qml.is_unitary` expects an `Operator` rather than an array/matrix, I couldn't see a way to perform the column-wise division, but if I understand correctly, to handle the case of unitarity up to a factor we should only need a single column norm? If the first column in $G$ has norm $|k|$, the first on-diagonal term of the product $G^\dagger G$ is $|k^2|$; if dividing by $|k|$ doesn't give a unitary matrix there can't be another constant which does (as $G^\dagger G \ne \mathbb{I}$). Let me know if there's any issue with this reasoning?

Reply:
Sure, I guess only the first norm is necessary. Then the coefficient is simply `np.linalg.norm(qml.matrix(gen)[0])`.